pipelinewise/fastsync/commons/target_snowflake.py

```diff
@@ -157,20 +164,33 @@ def copy_to_table(self, s3_key, target_schema, table_name, size_bytes, is_tempor
                 int(skip_csv_header),
             )
         else:
-            sql = """COPY INTO {}."{}" FROM 's3://{}/{}'
-                CREDENTIALS = (aws_key_id='{}' aws_secret_key='{}')
-                FILE_FORMAT = (type='CSV' escape='\\x1e' escape_unenclosed_field='\\x1e' 
-                field_optionally_enclosed_by='\"' skip_header={} COMPRESSION='GZIP' BINARY_FORMAT='HEX')
-            """.format(
-                target_schema,
-                target_table.upper(),
-                bucket,
-                s3_key,
-                aws_access_key_id,
-                aws_secret_access_key,
-                int(skip_csv_header),
-            )
-
+            if self.use_iam_accesskeys:
+                aws_access_key_id = self.connection_config['aws_access_key_id']
+                aws_secret_access_key = self.connection_config['aws_secret_access_key']
+                sql = """COPY INTO {}."{}" FROM 's3://{}/{}'
+                    CREDENTIALS = (aws_key_id='{}' aws_secret_key='{}')
+                    FILE_FORMAT = (type='CSV' escape='\\x1e' escape_unenclosed_field='\\x1e' 
+                    field_optionally_enclosed_by='\"' skip_header={} COMPRESSION='GZIP' BINARY_FORMAT='HEX')
+                """.format(
+                    target_schema,
+                    target_table.upper(),
+                    bucket,
+                    s3_key,
+                    aws_access_key_id,
+                    aws_secret_access_key,
+                    int(skip_csv_header),
+                )
+            else:
+                sql = """COPY INTO {}."{}" FROM 's3://{}/{}'
+                    FILE_FORMAT = (type='CSV' escape='\\x1e' escape_unenclosed_field='\\x1e' 
+                    field_optionally_enclosed_by='\"' skip_header={} COMPRESSION='GZIP' BINARY_FORMAT='HEX')
+                    """.format(
+                        target_schema,
+                        target_table.upper(),
+                        bucket,
+                        s3_key,
+                        int(skip_csv_header),
+                    )
```

Comment:
I also noticed that this part can be significantly shorter and actually we can do everything in one COPY command similar if we follow the relevant part in pipelinewise-target-snowflake [here](https://github.com/transferwise/pipelinewise-target-snowflake/blob/a26fa575cd24bdf63603d845988db90fbc0c5229/target_snowflake/db_sync.py#L470) .

Creating a "name external stage object on s3" in snowflake is mandatory, so we can do everything by one SQL:

```
        sql = """COPY INTO {}."{}" FROM @{}/{}
            FILE_FORMAT = (type='CSV' escape='\\x1e' escape_unenclosed_field='\\x1e' 
            field_optionally_enclosed_by='\"' skip_header={} COMPRESSION='GZIP' BINARY_FORMAT='HEX') 
        """.format(
            target_schema,
            target_table.upper(),
            self.connection_config['stage'],
            s3_key,
            int(skip_csv_header),
        )
```
We don't need to pass credentials to s3 because it's ready to use in s3.

Reply:
In our setup, we were creating external stage without any credentials since fastsync always passed the keys while loading the data and credentials were not mandatory param while creating external object . Your current suggestion would affect this setup, Do you think it is still fine to go ahead to pass the external stage alone ?

Reply:
You're right, your setup works with the current fastsync code but it's not really an expected behaviour. If you don't have an external stage to S3 with the correct credentials in snowflake then the first INCREMENTAL/LOG_BASED after fastsync will fail. INCREMENTAL and LOG_BASED are are using the [pipelinewise-target-snowflake connector](https://github.com/transferwise/pipelinewise-target-snowflake) and that requires the external stage. We should keep that behaviour here in fastsync as well.

So yes, please create an external stage for your s3 bucket and set the credentials in snowflake or use SF storage integration. Here we should use s3 directly only to upload the csv files.

Reply:
Thanks for the explanation, Adapted the fastsync as per your recommendation. Will try to update the documentation in the coming days :)

Reply:
hey @mashanm the related documentation updated by 5548489c73cf2998396c0f2e9e0d4e273a2cc7b4